### PR TITLE
Fix camera's unexpected transition when flying at negative height

### LINKED ISSUE #8610: Camera produces unexpected transition at negative heights
Here's a [Sandcastle](https://sandcastle.cesium.com/index.html#c=tVVdb9s2FP0rhF/stB5N8pIU2TpBsQzdS4EOSLC96EWVmEyoIhoU5dYt8t93JflDyrymHTa+iPfcT/IcUNsskG3pPrlALkntPpFr15TtA/29xxbzvDevfR2zsnZhviRf05rgii4EhH4LflsWLrw6JObBZdH94UNV3A4hi4u0frx4ndZpfZPVRZ41sXI0K4pf3F3WVvHW++pDFn5uY/T1Yv622pHoycY3ZSx9TTj2vGvrvDMWF4f2w8w0zx5cyOhdtbv1i72rW4VrYllnfYFX43NdZyHiLqthISVIJjgFrhVjApIlAa64MJxayaWwhncQcKO4otYIxQ1a+mJ56uND6fBq9n1GA/RDlMHl3xzhJ0a1hq6ZtAyEZHZJOswoK5QEoQVXzCwJoziNTRJrMFJZkGo8RLfazT82YRQUU8oyBdomEvvoriLThhmWWAlgAHeqA7E008IIboFpaSZtHkf7og2HQ8OA9gyfpfkZfsX/yC8YIyhIKRVAwgd+QSaUq0QzUGD3/FqgRlhuBTf/LbuMWSGQV1SOAMa4lj3B1ibWSgFKK/wkpgfx4rWw0lgmWAKg7A9xbLgGACa1EoBigaGkBG00NwlYVJfo5MWoSlBd2uquleBa/zuOVy8OpJF7F697nt72NJ04xLdlIBDflgmhr08BRx1c7mPpARkFne76GHWERmF4QUd/u+nHnM5xE0NZ32PQ/GuK3vTHBDUnL4/j0s9ooHIn2O4M9qXD8Iqn7Z4qa+r9Pnl1bY5R43lO4O4ceH6iZwTWlWg34y5o7SbW+bqPT4GJrE6ux/meyOBiG+oJYZ3qnnlVfnWRDCIk+FhEf/ZRyX3deEyv/P3iqWj30n6xmi1n6ybuKnc1JL0pHzY+RNKGakHpKrqHTYV/uGb1oc0/ukjzpuly16tD0root6QsLtPZk59nOiN5lTUNeu7aqropv7h0drVeYfwkrfJZgad+v3WhynZdyJ/86t0AUkrXKzT/nhWHCxlV/As) showing this issue around the Dead Sea, where the terrain has negative height. Try clicking "fly to position 2". Notice the camera will fly high up in the air, rotate slightly, then fly back down.

The expect transition here is for the camera to just rotate slightly and not do this extraneous transition.

This only seems to happen when doing camera flights at negative heights.

### PR DESCRIPTION
**Pull request by baothientran**
Fixes #8610

I added `maximumStraightFlyRadius` option to function `Camera.flyTo()`. The option specify the radius of an imaginary sphere with the destination point as a center. If camera is inside this sphere, it will fly straight to the destination, without doing any upside down parabolic curve. I think this option should satisfy the requirement we have for the camera: 
- fly straight to the nearest point (maximumStraightFlyRadius option is where user defines how far is near)
- fly straight if the `maxHeight > max(height(start), height(destination))` (this already exists)
- fly along the upside down parabolic curve if the two points are very far away from each other (even if they have negative height)

This is the [sandcastle's example](http://localhost:8080/Apps/Sandcastle/index.html#c=tZVNj9s2EIb/CuGL7dSlSQ5JkRvvougW6aVAg+yiveiiSFxHqEwZFOXECfa/dyR/ya7TTYBWPph8OZx5xXlAbbJANqX76AK5Jd59JPeuKdsV/aPXJuO8n97XPmald2E8I19ST/CJLgSU3oZ6UxYu3Bw25sFl0f1Zh6p43IVMpql/nr5Offd7yHyRZ02sHM2K4hf3lLVVfKzr6n0Wfm5jrP1k/KbakliTdd2Usaw94Vj1qfV5N5lMDwZ2rmmerVzI6FO1fawn+6XuKVwTS5/1CW6Gb3afhYijzMNESpBMcApcK8YEJDMCXHFhOLWSS2EN7yTgRnFFrRGKG5zp6exUpw6lw8PZ1xkY6E2UweX/auFHRrWGrpi0DIRkdkY6zSgrlAShBVfMzAij6MYmiTUYqSxINTTRPe36q0UYBcWUskyBtonEOrrLyLRhhiVWAhjAkepETM20MIJbYFqaszLPg3HRhsNLw0BeZZ/KVbt6iNj55YeIrXyXFWXb3BDJGNsH9jR8BYkXWBD/IwtgjKAgpVQACd+xADKhXCWagQK7Z8ECNcJyK7j5b0lgzAqBDCBlAhjjWvYwWJtYKwUorfAvMb2ITdLCSmOZYAmAst/Fg+EaAJjUSgCCBbuUErTR3CRgkUTRocioSpBEbXVXSnCtv4WHa11O/fxV6g+tI0sX7/tuvembdeok3ke7NuJ9dNbW16eAIw23+1h6UAZBpxM/Rh2lQRge03G9Xfdmz30gyqVfYtD4S4qr6fdhNSY/HO3STzhBfs+07RXtc6fhQZ+Xu+TrfPXbIOvKHKOGfk7i9pp43dELmHUp2vWwCs62Z7PreZ8vhTO4TkvP430jg4tt8GcN69h74W751UWyg5DglRHrq1dLXvumxu1VvZxcQnsE/NV8NBstmrit3N3B9U/lal2HSNpQTSidR7daV/hlbObv2/wvF2neNNO9/cV8uHVRlBtSFrfp6OLjm45IXmVNgytPbVU9lJ9dOrpbzDH+H1urGq9dv/x940KVbbuwD/zut51IKV3McXp9Z9wd0kXmvwE) for the option

This should make the camera's flight path more predictable. For example, if we want the camera to flight straight from one point to the other, we can put them within the radius of each other.

I also add @dennisadams 's [comment link](https://github.com/CesiumGS/cesium/issues/8610#issuecomment-656137112) to the getAltitude() function as it is a very reasonable explanation for the math behind the function.

@lilleyse @OmarShehata Can you please take a look at it? Please let me know if this is what we expect the camera to behave

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @baothientran!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/camera-transition/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by baothientran** [comment]
@lilleyse for the default value of the maximum radius, I just place the camera in different position around the destination and console.log() the distance between them to choose a reasonable distance. There is no math behind it though. I will provide the sandcastle example and update your PR comments soon

**Comment by baothientran** [comment]
@lilleyse I updated the PR to address the reviews and created the [Sandcastle's example](http://localhost:8080/Apps/Sandcastle/index.html#c=tVZbj+I2GP0r1rwAW2p8iR17buqWdvuyakedUatKSKtsYpioJkGOw5Zd8d/7OSEhUKbbrVpeyHf/fM6JYZs4tM3NB+PQHSrMBzQ3VV6v8S+NbzxKG3NeFj7JC+NGU/RpUSD4eOMcuB5cuc0z4667wtSZxJtfS2ezpzZlPFkU+8nNolgUW5i2KfPCz0trTerzsoCx7XxcpaYweOPyde7zralwkmXjwUoPofChCx87jCeheWidJmvjkjc2Xz37h8Q/Q+9lYisTwqczzM+mgK3DiO+3pvBv88pDxI2XddE2bTInzWFD66WD3o8ejgZN2y7vjr7mcAjlSzQ+32HSARaCh5NkZglgZu0M3Fbgd2ntHKzSlk76OnSOWINLHwzhKg/+a3TSr3NPj5lpacsjU/Ng4flvr3/sUxqamofwtUcGwOv2ON/CmXW5Na+tHbdF+57lx6TI0qTy1oRVn8rVyppva+8B1ZFvLORL5BoG0AYwAlWdwzZFPRPbxNbmAMcFhpvwzQvDvzPLpLb+qSzt+8R1S7yxu7BBhxCio8G4HviDZA5wLu3uqRzCnpnK50XSNLgevjrzxHl4Sgo+jiIeEUYxp1IQwng8RZwKyhTFOqIR04oGF6dKUIG1YoIqsORkQFrpclBFN2ewQLNE7g6v0UsrfE2wlDwMizThLCJ6ioJPCc1ExJlkVBA1RQTDNjqOtYJMoXkkhkuET715cQjBXBAhNBFc6jiCOTJ0JFIRRWIdca44PInghNZEMsWo5kRG6mTMfvCc1a47ND+IcnLzosb+ll/2P/LLlWKYR1EkOI9pyy+PYkxFLAkXXB/41RwrpqlmVP237BKiGQNeQTmME0Jl1BCsdax1xLiQAr5i1TgBeMl0pDRhJOZc6C/iWFHJOSeRFIyDWHjbMuJSSapirkFdLMiLYBGDuqSWYRSjUv47jmevOtLQyvj54dUPNB057G/94+9Ia98cE3od3KGzu3GQdMS6z+pdgzQAqI/Xm8O9P9zj0bu8WEHS6NMCoosvE9QIfdWvi/8AA5R74ttd8H0MPoD4dNy5sk6j/0xeYUyfNdzn6Nxdcl7e6DMCCy3qzXAKWLsT63Lf/bnjRFbH0H50INIZX7vihLCgus/cKj8Yj1oRIrgsfHnxUknLoiqh3Jar8bloD9J+NbuaXt1WfmfNfbfzN/l6UzqPamfHGM+8WW8s/K+oZu/r9HfjcVpV3U/z7WxYepvlW5Rnd4urs39qiyuU2qSqILKsrX3MP5rF1f3tDPL/UmrLJAMMftoaZ5NdSHum929bJ8b4dgbm5UrfQnTW+U8) to render flight path. Please let me know if I need to add anything.

File: Source/Scene/CameraFlightPath.js [hunk 1]
```diff
@@ -20,6 +20,8 @@ import SceneMode from "./SceneMode.js";
 var CameraFlightPath = {};
 
 function getAltitude(frustum, dx, dy) {
+  // More details about the math behind this function is in
+  // https://github.com/CesiumGS/cesium/issues/8610#issuecomment-656137112
```
[the remark this answers is not in the document]
Comment (baothientran) [reply Source/Scene/CameraFlightPath.js:24]: wait me for a second. I need to update this one

**Comment by mramato** [comment]
Sorry if I'm confused, but does this PR actually fix the problem, or does it add a new parameter that has to be set by the end user in order to fix the problem?

**Comment by baothientran** [comment]
@lilleyse updated

**Comment by baothientran** [comment]
@mramato Currently, the flight path of the camera is based on the function `getAltitude()` in CameraFlightPath.js file. It approximates the altitude to let the camera see the destination point. The rule is that if the `getAltitude() > max(height(destination), height(start))`, the camera will flight along the upside down parabolic curve. Otherwise, it will flight straight to the destination. 

One of the problem with the function is that it only returns positive height. If `max(height(destination), height(start)) < 0`, the rule mentioned at the beginning will be true. As a result, the camera will flight along the upside down parabolic curve even if the two points are close together.

Another problem is that since it is an approximation and depends on the start height and destination height, I think there is no guarantee it will return the correct value to let the rule above to be false so that the camera will fly straight if the start and destination are close enough.

It would be ideal to have a better approximation for the `getAltitude()` function to address the problems above, but I can't find any.

I adds the option `maximumStraightFlightDistance` to address both of the problems above. If the distance between the camera and the destination is within `maximumStraightFlightDistance`, the camera will flight straight to it. This will be the main rule to determine the flight path of the camera, instead of using the height rule above. I think the distance rule make the camera's flight path more predictable than the height rule. Currently, it is defaulted to be 3000.0 meter. If the client wants to make the camera flight straight inside a building with radius  more than 3000.0 (I guess it is rare), they can set the distance accordingly to choose the flight path.

I'm not sure if the explanation addresses your question. Please let me know if you have any question about the solution

**Comment by baothientran** [comment]
@mramato I think I should have a more straightforward answer. it should fix the issue. The users don't have to set the new parameter as it is defaulted to be 3000.0 meter surrounding the destination point. Unless they have a more special case, they can set it themselves to choose the better flight path.

**Comment by mramato** [comment]
@baothientran sorry for being obtuse but you're saying it fixes the issue, but only as long as the camera is within 3000m of the destination?  Might be okay, I genuinely don't know.

As general advice about our approach to designing APIs, it's usually bad practice to add additional options for the user simply because we can't figure out how to solve the problem generically.  What we are actually doing is pushing the problem onto the user and they aren't going to know the answer either.  However, if the default value of 3000 works in most cases here, then this is probably the right decision and what we want to do.  I just didn't see much discussion about this from the end user/dev's point of view in the PR.

Super excited to see flights getting some love!  Thank you!

**Comment by baothientran** [comment]
@mramato I agree with the advice. Since this PR modifies the public API and requires user's assistance to have a correct flight path, I quite hesitate with it from the beginning, and it should be the last thing to do. It may have a backward compatibility issue if we happen to have a better generic solution in the future as well.

To get the discussions rolling, I think I should jot down some of my thoughts. 

For the camera flight, I think it should have the following requirements (they may not be true):

- If the start and destination points are close, it should fly a straight path. This way, the user may not lose the context of the surrounding. For example, let the camera fly from one point to the other in the underground tunnel, or fly inside the museum. This has a subsequent question about the near definition. does the user expect the camera to fly straight in a straight tunnel that is 6000 meters in length? Sometimes, the camera is in the mountains area, and the user doesn't want it to flight straight from one point to the other though their distance is within only 500 meters. The camera can pass through the terrain and user will lose the context of surrounding. I feel like having the upside down curve is more appropriate in this case.

- If the start and destination points are very far away(e.g it flies from Australia to the US), it should fly upward before going down to the destination (the upside down parabolic curve). If it flies near the surface of the earth without going upward, the camera can hit the terrain and the user can lose the context.

- Assume we have a solution that can automatically decide those two paths, the user should be able to correctly and intuitively predict the camera flight's path based on the parameter set in the function `Camera.flyTo()`

When I first tried to find an automatic solution without modifying the public API, I have a trouble with the 1st requirement as the near distance definition depends on different use cases. Currently, there are no options to let the user set the maximum distance between two points within which the camera will flight straights. If we end up with a solution that can let the user indirectly calculate the maximum distance based on the existing options in `Camera.flyTo()`, it will fail the 3rd requirement. It also implies that the user has to look at the implementation of the `flyTo()` function in this case.

With the option to let the user set the maximum straight flight distance, the user may not know the value initially. But they can  try different values to have the closest expected behavior.

@mramato I may overthink the problem, but I would love to have more opinions on it

**Comment by lilleyse** [comment]
Just weighing the tradeoffs more here, there's some situations where straight line distance is not desired like if the camera goes underneath terrain during its path. We just need to be careful about changing the defaults and making sure it doesn't hurt common cases like this.

flight
[Local sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=tVZrj+I2FP0rFl8AlZr4ESeeYUbd0m6/rNpRZ9SqEtIqmxjGqkmQ47DLrvjvvc6LQJl9SC0fZuJ7r+89PufEsE8s2mv1Xll0h3L1Hi1Vqast/qOOTcZpvVwWuUt0rux4hj6tcgQfp6yF0IMt9jpT9qbbmFqVOPVnYU321JRMpqv8OL1d5at8D9N2hc7dsjBGpU4XOYxt5uMyVbnCO6u32um9KnGSZZMBpAe/8aFLnzpMpr65b50mW2WT10Zvnt1D4p6h9zoxpfLp8xnqd5UDaj/i573K3RtdOsjYybrKm6Z15bQ+rG+9ttD70cHRoGnT5e0pVh8OIb1Gk0sM044wn2xPkqk1kJk1M3CzA79NK2sBSrN12u9Dl4zVvPRJny61j9+gs35deHaqTAtTnJRa+hVe/vXq176klql+8P+OSAF5HY5LFFZti716Zcyk2XTsVX5M8ixNSmeUh/pUbDZG/Vg5B6yOXb1CrkC2VgDtgCNw1SVtM9QrsU9MpVo6rihcp29fGP6TWieVcU9FYd4ltgPx2hw8go4hRMaDcT3xrWVaOtfm8FQMac9U6XSe1A1uhq/OMrEOnpKcTTgPBBUxZiKUhDNCZ4gyKkXAMA8JjQiTfIaYFBEVAY6IgCLKJJsORCusBld0cwYAahDatq/RSxACHPIAcDARCxgXxHKGvg8wiSJGQ8plSGNOWRTV0ZhSHvOABhAK4/AMh/9Uu8/MiQlMCKNAShnGUQAdA8yYEHEgSEAJD0LGfIyLkBFCIkLhb8CHM46D56yy3aFZa8rp7Yse+6y+9H/UN465wCKOaUwjzyLoGwvBMWeCR3BKzwMIzCVjmIVcUsGpjP5rgQPGgsgrKSNKOoFFTENQRDIuecQ7gUXEOSGCcsArvlVgCZrJUEqYSHgrMLgokpIGRAgaNwLz2FuBSC6DczN/vcCdXmij3LJ9671CJ/n6C//0FdKsb08FvQXu0MW1OCg6cdxX9aFBGXDT56tde+UPcTw6q/MNFI0/rSC7+jYvjdF3PVz8ARZg2rPY4Urso48BwefjLh11nv06W/kxfdUQzyl4uBa8jugL3vItqt1wCqwOZ6vrfY+XgTNTnVLHcSukVa6y+Zlg3nNfuFB+UQ41JkRwT7ji6n2SFnlZwHZTbCaXpm2NPZqNFqU7GHXfIf5Bb3eFdaiyZoLx3KntzsAPinL+rkr/Vg6nZdl9Jy/mw62LTO+Rzu5Wo4ufaKsRSk1SlpBZV8Y86o9qNbpfzKH+X1tNkWTAwG97ZU1y8GXP5P5NE8QYL+awvL7TNQRddP4H)

**Comment by baothientran** [comment]
@lilleyse I has the idea that from the application's perspective, it would be helpful to render the influence sphere around the destination point. It is intuitive to visualize which area the camera will fly straight to the destination and adjust the default setting accordingly.

**Comment by lilleyse** [comment]
@baothientran informed me offline that `master` actually has the same behavior as the gif I posted above and it's because the height difference between the two points isn't large enough to trigger the parabolic arc.

Here's a different Sandcastle that shows the problem better: [local](http://localhost:8080/Apps/Sandcastle/index.html#c=tVZbj+M0GP0r1ry0FcUT2/ElcxNLYXlZwYgZgZAqrbKJ241w48pxunRX/e98zq1p6bAsgjy08Xf3OSdOdqlDu0J/0A7do1J/QAtdFfUG/9LYppOsWS5s6dOi1G4yR5+WJYLLa+fA9Ojsrsi1u+kTM6dTr3+1zuTPbch0tiwPs9tluSx30G1ri9IvrDE684UtoW3bH1eZLjXeumJT+GKnK5zm+XQ00mNIfOzdxwrTWSgeSmfpRrv0tSnW7/1j6t9D7VVqKh3cpz30z7qEqUOL73e69G+KyoPHTVd12RZtImfNZkPplYPaTx62BkXbKm+PtmZzCBUrND2fYdYDFpzdTnK9AjDztgduM/DbrHYORmlTZ0MeOkeswWVwBndVBPsNOqnXm+fHyMwae2RqEVZ48durH4eQhqbmJvwdkAbw+jnOp3B6Y3f6lTHTNukwsPyUlnmWVt7oMOqzXa+N/rb2HlCd+GaFvEWuYQBtASNQ1TlsczQwsUtNrTs4LjDcuG9faP6dXqW18c/Wmnep64d4bfZhgh4hRCajdgPwnWQ6OFdm/2zHsOe68kWZNgVuxo/OInUe7tKSTeOYxUIqLGJOJY2iZI4YiRMlGI4SLnnEKQETI0ksYkwUU4pyRmYjzqwrQBR9m1H/ZobCdU/RSxN8HWHFKEkoYyKOFIffOYowjyVPYISY8DjhSs4RBEZKEqkY44yqiFBGx3OEq96+2CfCMU1oRJiUSRJxFcU89JGMC5WomAgZ0aRtTVkiSSJhp1TK0yaH0X1eu37XrBPl7PZFjf0tv/R/5FeoWGJOuKKxoi29AAJmESBJlZSi4Rd2K7AAMKSIFWH/LcEykgI4k0wkFBimAWXBJYlELBPFJIiqsVGeCKUIkSAxQsSX0SskiyOox3gsOAi4oZLwRAY500gmgnf0JqAEAeICEcXJv2O3JwuttV90j3yg58jdcNof3x/t+vYYMPB/j87OxFHQEeEhajCNwgCbwV9vu/N+PMeTd0W5hqDJpyV4l18mpAn6ahgX/wELUOyJbX/B9jHYAODTdud6OvX+M1GFNkPUeJ6jcX/JeHmiz2grlKi34y6w2p+sLtc9nBtORHV0HSYdkU772pUnhAXNfeY0+UF71IoQwSHh7cXDJLNlZSHd2PX0XLSdsK/mV3eV3xv90E/8TbHZWudR7cwU42uvN1sDXxPV9bs6+117nFVV/0K+ux6n3uXFDhX5/fLq7PtseYUyk1YVeFa1MU/FR728eri7hvi/pBqb5oDATzvtTLoPYe/Jw5vWiDG+u4bl5UzfAnRW+U8) vs. [master](https://sandcastle.cesium.com/#c=tVZbj+M0GP0r1ry0FcUT2/ElcxNLYXlZwYgZgZAqrbKJ241w48pxunRX/e98zq1p6bAsgjy08Xf3OSdOdqlDu0J/0A7do1J/QAtdFfUG/9LYppOsWS5s6dOi1G4yR5+WJYLLa+fA9Ojsrsi1u+kTM6dTr3+1zuTPbch0tiwPs9tluSx30G1ri9IvrDE684UtoW3bH1eZLjXeumJT+GKnK5zm+XQ00mNIfOzdxwrTWSgeSmfpRrv0tSnW7/1j6t9D7VVqKh3cpz30z7qEqUOL73e69G+KyoPHTVd12RZtImfNZkPplYPaTx62BkXbKm+PtmZzCBUrND2fYdYDFpzdTnK9AjDztgduM/DbrHYORmlTZ0MeOkeswWVwBndVBPsNOqnXm+fHyMwae2RqEVZ48durH4eQhqbmJvwdkAbw+jnOp3B6Y3f6lTHTNukwsPyUlnmWVt7oMOqzXa+N/rb2HlCd+GaFvEWuYQBtASNQ1TlsczQwsUtNrTs4LjDcuG9faP6dXqW18c/Wmnep64d4bfZhgh4hRCajdgPwnWQ6OFdm/2zHsOe68kWZNgVuxo/OInUe7tKSTeOYxUIqLGJOJY2iZI4YiRMlGI4SLnnEKQETI0ksYkwUU4pyRmYjzqwrQBR9m1H/ZobCdU/RSxN8HWHFKEkoYyKOFIffOYowjyVPYISY8DjhSs4RBEZKEqkY44yqiFBGx3OEq96+2CfCMU1oRJiUSRJxFcU89JGMC5WomAgZ0aRtTVkiSSJhp1TK0yaH0X1eu37XrBPl7PZFjf0tv/R/5FeoWGJOuKKxoi29AAJmESBJlZSi4Rd2K7AAMKSIFWH/LcEykgI4k0wkFBimAWXBJYlELBPFJIiqsVGeCKUIkSAxQsSX0SskiyOox3gsOAi4oZLwRAY500gmgnf0JqAEAeICEcXJv2O3JwuttV90j3yg58jdcNof3x/t+vYYMPB/j87OxFHQEeEhajCNwgCbwV9vu/N+PMeTd0W5hqDJpyV4l18mpAn6ahgX/wELUOyJbX/B9jHYAODTdud6OvX+M1GFNkPUeJ6jcX/JeHmiz2grlKi34y6w2p+sLtc9nBtORHV0HSYdkU772pUnhAXNfeY0+UF71IoQwSHh7cXDJLNlZSHd2PX0XLSdsK/mV3eV3xv90E/8TbHZWudR7cwU42uvN1sDXxPV9bs6+117nFVV/0K+ux6n3uXFDhX5/fLq7PtseYUyk1YVeFa1MU/FR728eri7hvi/pBqb5oDATzvtTLoPYe/Jw5vWiDG+u4bl5UzfAnRW+U8)

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by OmarShehata** [comment]
I had a chance to test this out a bit - my opinion is we should not expose a new parameter and instead hard-code a reasonably small default distance to force a straight flight. 

@lilleyse the issue you describe in https://github.com/CesiumGS/cesium/pull/9025#issuecomment-660258357 of this causing the camera to go through the terrain in some cases I think will not be worse by this change. In general, the CesiumJS camera does not check for collisions when making a flight (right?). So there will always be a way to construct a flight where it goes through terrain/buildings. 

I actually did not realize that setting `maximumHeight` can solve the original issue from the user's side. I think it is reasonable that this value gets computed based on the flight destination/distance. Very short flights get a low maximumHeight that triggers a straight flight. The user can override this by explicitly passing a value.

This seems a lot simpler than thinking about both a maximum height and the threshold at which the maximum height is ignored. From the user's perspective, the choice is either use the default, which is computed dynamically, or pass your own. 

Let me know what you think and I can make this change if Bao doesn't have the bandwidth.

**Comment by lilleyse** [comment]
@OmarShehata It's fine if you take over. @baothientran what's your thoughts on just using `maximumHeight`?

**Comment by baothientran** [comment]
@lilleyse @OmarShehata if we are able to fix the issue without adding new option to `flyTo()` (which already has many), I'm in favor of it. 

I have a question. Currently the `maximumHeight` is generated based on the flight distance, but the formula doesn't work well with negative height. Are we able to find a way to fix the math?

**Comment by OmarShehata** [comment]
My suggestion was just to keep the behavior in this PR but not expose that new parameter (so it'll be a private variable set to the default 3000 or whatever we decide). I didn't realize that didn't work well with negative heights? I can test this when I get a chance to look at this again.

**Comment by baothientran** [comment]
@OmarShehata That makes sense to me now. The PR threshold distance (currently default to be 3000) should allow us to bypass the `maximumHeight` 's current formula at the negative height, so it works well in that case. I also tested different locations in the past but didn't find any change in other existing behaviors

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @baothientran! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by ggetz** [comment]
@baothientran To keep things tidy, I'm going to close this PR. If you end up getting some time to work on this, feel free to reopen. Thanks!

**Pull request closed without merging**